### PR TITLE
Technical Guidelines: domain vs presentation-specific logic

### PR DESCRIPTION
**Pull request by buskamuza**
## This PR is a:

- [ ] New topic
- [x] Content update
- [ ] Content fix or rewrite
- [ ] Bug fix or improvement

## Summary

> Configuration for the presentation layer SHOULD be declared in corresponding application area
Reason: so that the configuration impacts only necessary areas. "SHOULD", so that intentional declaration in `global` area is permitted.
> Domain/business logic MUST be executed on the Service Contracts layer
Very important topic! Domain-specific logic MUST never be executed in presentation layer (so should be executed on Service Contracts layer or below). Violation of this rule may cause several issues:
1. Limit reusability of such code. Example: calculating price in a block limits reuse of the same algorithm in a CLI application. Or relying on session data (which can be used in presentation layer only, see 6.2.2.), also limits the logic to the presentation layer only and does not allow to reuse it in a CLI application.
2. Incorrect customizations. If a customization for a domain logic (e.g., creating a product) is applied on presentation layer (e.g., in `adminhtml` area only), same customization will not happen when somebody will implement a CLI command for adding a product. As those two implementations are done by different developers (including extension developers), it may be hard to catch the issue. And to fix it, the logic will have to be moved to the Service Layer anyways.

## Additional information

Affected URLs:

- https://devdocs.magento.com/guides/v2.0/coding-standards/technical-guidelines.html
- https://devdocs.magento.com/guides/v2.1/coding-standards/technical-guidelines.html
- https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html
- https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html

whatsnew
Added information about domain- versus presentation-specific logic to the [Technical Guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html).

### CODE REVIEW

**Comment by buskamuza** [comment]
❗️ Please wait for architectural approval before merging.

**Comment by magento-cicd2** [comment]
Can one of the admins verify this patch?

**Comment by lorikrell** [comment]
Happy to review when ready!

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 1]
```diff
@@ -514,6 +514,10 @@ class View extends Template
 
 6.2.4. Actions MUST NOT reference blocks declared in layout.
 
+6.2.5. Configuration for the presentation layer SHOULD be declared in corresponding application area. This includes events and plugins that customize presentation layer
+
+6.2.5.1. Configuration that applies to all areas CAN be declared in `global` area
```
Comment (buskamuza) [inline guides/v2.0/coding-standards/technical-guidelines.md]: Is there a use case for this?

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 2]
```diff
@@ -528,6 +532,10 @@ class View extends Template
 
 We are reviewing this section and will publish it soon.
 
+6.4.1. Domain/business logic MUST be executed on the Service Contracts layer
```
Comment (buskamuza) [inline guides/v2.0/coding-standards/technical-guidelines.md]: Should we add "Domain layer"? Consider it when adding "Service Layer" section. These rules can be moved there then.

File: guides/v2.0/coding-standards/technical-guidelines.md (hunk 1 above)
Comment (buskamuza) [reply guides/v2.0/coding-standards/technical-guidelines.md]: Didn't find a good use case for this. Removed and made 6.2.5 more strict.

**Comment by jeff-matthews** [comment]
@buskamuza, @paliarush approved. Can we merge?

**Comment by magento-cicd2** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 3]
```diff
@@ -525,7 +527,81 @@ class View extends Template
 
 ### 6.4. Service Contracts (Application) layer
 
-We are reviewing this section and will publish it soon.
+6.4.1. Location of API interfaces
+
+6.4.1.1. Service contract interfaces SHOULD be placed in separate API modules. The other modules will depend on the API module, and implementations could be easily swapped via `di.xml`. API module namess must end with the `Api` suffix. For example, if a module is named `MyModule`, its APIs SHOULD be declared in a module named `MyModuleApi`.
```
Comment (kandy) [inline guides/v2.0/coding-standards/technical-guidelines.md]: > API module names**s** _must_ end with the `Api` suffix

I think that there should be "SHOULD" not "must"

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 4]
```diff
@@ -525,7 +527,81 @@ class View extends Template
...
+
+* An array of items with retriable errors
+
+* An array of items with non-retriable errors
+
+6.4.3.5. Batch operations that modify state SHOULD be implemented in the most performant manner and SHOULD NOT load modified entities to generate response. 
+
+6.4.3.6. Asynchronous invocation of the command services SHOULD be supported by the web API framework.
+
+6.4.3.7. Operation UUID MAY be provided by the client during service invocation. UUID MUST allow the client to get the operation status information.
```
Comment (kandy) [inline guides/v2.0/coding-standards/technical-guidelines.md]: Better to replace UUID with just ID

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 5]
```diff
@@ -525,7 +527,81 @@ class View extends Template
...
+
+6.4.3.8. Data objects returned by service contracts SHOULD be fully loaded to ensure consistency.
```
Comment (kandy) [inline guides/v2.0/coding-standards/technical-guidelines.md]: I propose to add here "Related entities should be not loaded"

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 6]
```diff
@@ -525,7 +527,81 @@ class View extends Template
...
+
+6.4.4. Service Implementation
+
+6.4.4.1. Service data interfaces SHOULD extend from `Magento\Framework\Api\ExtensibleDataInterface`. The only exception is when extensibility is not desired, such as in case of value-objects.
```
Comment (kandy) [inline guides/v2.0/coding-standards/technical-guidelines.md]: SHOULD extend from -> SHOULD implements

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 7]
```diff
@@ -525,7 +527,81 @@ class View extends Template
...
+
+6.4.4.2. Extensible data interfaces MUST NOT form hierarchies. If interface `MyInterface` extends `ExtensibleDataInterface`, there must be no interfaces extending `MyInterface`. Otherwise, a list of extension attributes will be shared for all extensible interfaces in the hierarchy.
+
+6.4.4.3. Service implementations and plugins MUST NOT rely on storage-specific integrity features, such as foreign key constraints.
```
Comment (kandy) [inline guides/v2.0/coding-standards/technical-guidelines.md]: Better to change: Service implementations and plugins -> Extension of Services from other modules

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 8]
```diff
@@ -525,7 +527,81 @@ class View extends Template
...
+
+6.4.4.4. Replacement strategy SHOULD be used to persist main entity fields/attributes, child entities, and relation links.
```
Comment (kandy) [inline guides/v2.0/coding-standards/technical-guidelines.md]: 6.4.4.4 and 6.4.4.5 better replace with:  Persist of main entity fields/attributes, child entities, and relation links MUST be *idempotent*

File: guides/v2.0/coding-standards/technical-guidelines.md (hunk 3 above)
Comment (paliarush) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
6.4.1.1. Service contract interfaces SHOULD be placed in separate API modules. The other modules will depend on the API module, and implementations could be easily swapped via `di.xml`. API module names must end with the `Api` suffix. For example, if a module is named `MyModule`, its APIs SHOULD be declared in a module named `MyModuleApi`.
```

Comment (paliarush) [reply guides/v2.0/coding-standards/technical-guidelines.md]: What is the case when we don't want this?

File: guides/v2.0/coding-standards/technical-guidelines.md (hunk 6 above)
Comment (paliarush) [reply guides/v2.0/coding-standards/technical-guidelines.md]: One interface should extend another, so I think current statement is correct.

File: guides/v2.0/coding-standards/technical-guidelines.md (hunk 7 above)
Comment (paliarush) [reply guides/v2.0/coding-standards/technical-guidelines.md]: How about `Service implementations and extensions MUST NOT...`

File: guides/v2.0/coding-standards/technical-guidelines.md (hunk 8 above)
Comment (paliarush) [reply guides/v2.0/coding-standards/technical-guidelines.md]: I would keep 6.4.4.5, 
6.4.4.4 can be changed to have `idempotency` requirement

**Comment by buskamuza** [comment]
@jeff-matthews , this one can be merged now.

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 9]
```diff
@@ -513,6 +513,8 @@ class View extends Template
 
 6.2.4. Actions MUST NOT reference blocks declared in layout.
 
+6.2.5. Configuration for the presentation layer MUST be declared in corresponding application area. This includes events and plugins that customize presentation layer
```
Comment (jeff-matthews) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
6.2.5. Configuration for the presentation layer MUST be declared in the corresponding application area. This includes events and plugins that customize the presentation layer.
```

File: guides/v2.1/coding-standards/technical-guidelines.md [hunk 10]
```diff
@@ -514,6 +514,8 @@ class View extends Template
 
 6.2.4. Actions MUST NOT reference blocks declared in layout.
 
+6.2.5. Configuration for the presentation layer MUST be declared in corresponding application area. This includes events and plugins that customize presentation layer
```
Comment (jeff-matthews) [inline guides/v2.1/coding-standards/technical-guidelines.md]:
```suggestion
6.2.5. Configuration for the presentation layer MUST be declared in the corresponding application area. This includes events and plugins that customize the presentation layer.
```

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 11]
```diff
@@ -506,6 +506,8 @@ class View extends Template
 
 6.1.2. Every application layer (Presentation, Service Contracts, Data Access) MUST process (handle or re-throw) exceptions of the underlying layer.
 
+6.2.5. Configuration for the presentation layer MUST be declared in corresponding application area. This includes events and plugins that customize presentation layer
```
Comment (jeff-matthews) [inline guides/v2.2/coding-standards/technical-guidelines.md]:
```suggestion
6.2.5. Configuration for the presentation layer MUST be declared in the corresponding application area. This includes events and plugins that customize the presentation layer.
```

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 12]
```diff
@@ -604,6 +606,10 @@ class View extends Template
 
 6.4.4.10. Service data interfaces MUST NOT contain any business logic. They SHOULD represent a container of data that is transferable over the wire. All the business logic SHOULD be moved to services.
 
+6.4.4.11. Domain/business logic MUST be executed on the service contracts layer
```
Comment (jeff-matthews) [inline guides/v2.2/coding-standards/technical-guidelines.md]:
```suggestion
6.4.4.11. Domain/business logic MUST be executed on the service contracts layer.
```

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 13]
```diff
@@ -604,6 +606,10 @@ class View extends Template
...
+
+6.4.4.12. Any customizations to the domain/business logic MUST be executed on the Service Contracts layer, and so MUST be declared in `global` area of configuration
```
Comment (jeff-matthews) [inline guides/v2.2/coding-standards/technical-guidelines.md]:
```suggestion
6.4.4.12. Any customizations to the domain/business logic MUST be executed on the Service Contracts layer, and so MUST be declared in the `global` area of configuration.
```

**Comment by jeff-matthews** [comment]
running tests

**Pull request merged**